include/envoy/network/io_handle.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include <memory>
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle
+ */
+class IoHandle {
+public:
+  IoHandle(int fd=-1):fd_(fd) { }
+  virtual ~IoHandle() { }
+
+  int fd() const { return fd_; }
+
+  // implicit cast operator
+  operator int() const { return fd_; }
+
+  // Assignment operators
+  void operator = (int fd) { fd_ = fd; }
+  void operator = (const IoHandle& ioHandle) { fd_ = ioHandle.fd(); }
```

Comment:
I'd prefer we nix equality and explicit check io->fd() just to make it clear where we need to remove fd access.
I'd hope we could get away without an assignment operator but let's see what happens when we split the interface from the impl :-)

Reply:
@alyssawilk, @mattklein123 So without checking in new code yet, is this closer to what you had in mind:

io_handle.h:

class IoHandle {
public:
  IoHandle() { }

  virtual ~IoHandle() { }

  // TODO(sbelair2) remove fd() method
  virtual int fd() const PURE;
};

And then in source/common/network, a new file io_socket_handle_impl.h:

/**
 * IoHandle derivative for sockets
 */
class IoSocketHandle : public IoHandle {
public:
  IoSocketHandle(int fd=-1):fd_(fd) { }
  virtual ~IoSocketHandle() { }

  // TODO(sbelair2) remove fd() method 
  int fd() const override { return fd_; }

private:
  int fd_;
};
typedef std::unique_ptr<IoSocketHandle> IoSocketHandlePtr;

@alyssawilk: I prefer IoSocketHandle, IoVppHandle to the derivatives you named above- what do you think about that? 

I can make the necessary changes for these new class definitions, but I'd like to know if you agree first. So sorry for all the churn!

Reply:
Yup, that's exactly right.

Reply:
+1 on interface.  No preference on naming - I'm terrible at naming :-P

Reply:
@alyssawilk, @mattklein123 I have a question for you. In this PR, how far do you want to go in the refactoring of the fd with respect to the calls which take an fd. I.e., the interface to these calls will eventually change from, for example:

Api::SysCallIntResult Ipv4Instance::bind(int fd) to something like
Api::SysCallIntResult Ipv4Instance::bind(IoHandle* io_handle)

My question is, in the interest of reducing more churn, which would you prefer for this immediate PR? Thank you :-)

Reply:
Ultimately we are going to go to `io_handle.bind(address)`, right? So then maybe just the former for now is fine?